gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/actors/CheckPermissionsActor.java

```diff
@@ -0,0 +1,398 @@
+     *
+     * @param command the {@link CheckPermissions} command.
+     * @return a map grouped by entity ID, containing permission check wrappers.
+     */
+    private Map<String, Map<String, ImmutablePermissionCheck>> groupByEntityAndPolicyResource(
+            final CheckPermissions command) {
+        return command.getPermissionChecks().entrySet().stream()
+                .collect(Collectors.groupingBy(
+                        entry -> entry.getValue().getEntityId(),
+                        Collectors.toMap(
+                                Map.Entry::getKey,
+                                entry -> entry.getValue()
+                        )
+                ));
+    }
+
+    private CompletableFuture<Map<String, PermissionCheckWrapper>> retrieveOrSetPolicyId(
+            final String entityId,
+            final Map<String, ImmutablePermissionCheck> permissionCheckMap,
+            CheckPermissions command) {
+
+        if (permissionCheckMap.values().stream().findFirst().get().isPolicyResource()) {
+            return CompletableFuture.completedFuture(
+                    permissionCheckMap.entrySet().stream()
+                            .collect(Collectors.toMap(
+                                    Map.Entry::getKey,
+                                    entry -> new PermissionCheckWrapper(entry.getValue(), PolicyId.of(entityId))
+                            ))
+            );
+        }
+
+        return retrievePolicyIdForEntity(entityId, command.getDittoHeaders())
+                .thenApply(policyId ->
+                        permissionCheckMap.entrySet().stream()
+                                .collect(Collectors.toMap(
+                                        (Map.Entry<String, ImmutablePermissionCheck> entry) -> entry.getKey(),
+                                        entry -> new PermissionCheckWrapper(entry.getValue(), policyId)
+                                ))
+                ).toCompletableFuture();
+
+    }
+
+
+    /**
+     * Handles the completion of policy retrieval by checking the permissions for each policy.
+     *
+     * @param command the {@link CheckPermissions} command.
+     * @param permissionResults the map to store the permission check results.
+     * @param enrichedGroupedByEntityId the permission checks grouped by entity ID and enriched with policyId.
+     */
+    private CompletionStage<Void> handlePolicyRetrievalCompletion(final CheckPermissions command,
+            final Map<String, Boolean> permissionResults,
+            final Map<String, Map<String, PermissionCheckWrapper>> enrichedGroupedByEntityId) {
+        final Map<PolicyId, Map<String, ResourcePermissions>> aggregatedPermissions =
+                groupPermissionsByPolicyId(enrichedGroupedByEntityId);
+
+        final List<CompletionStage<Void>> permissionCheckFutures = aggregatedPermissions.entrySet().stream()
+                .map(aggregateEntry -> checkPermissionsForPolicy(aggregateEntry, command, permissionResults))
```

Comment:
@hu-ahmed I had a closer look here .. this is a potential concurrency issue - as you modify the (non-concurrent) `permissionResults` Map as "side effect" in `checkPermissionsForPolicy`.
But we also should not "just" use `ConcurrentMap` - instead change it to a functional style.

To change that to a functional style, `checkPermissionsForPolicy` should instead return a CompletableFuture of an immutable map of calculated permission results.

And in this method we would "combine"/merge all of the calculated maps to a combined map.

Sorry I didn't see that before.